### PR TITLE
Category

### PR DESCRIPTION
**Pull request by VlasenkoYevhenii**

### CODE REVIEW

**Review by ReamFOX — APPROVED** [review]
LGTM. Just a tiny comment. And you probably have here too many changes not related to the straight to the task.

File: src/main/java/com/example/bookstoreapplication/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+@RequestMapping("/categories")
+@Tag(name = "Categories", description = "Category management")
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PostMapping
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @Operation(summary = "Category creation",
+                description = "Add new category to the DB(ADMIN only")
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapplication/controller/CategoryController.java:34]:
```suggestion
                description = "Add new category to the DB(ADMIN only)")
```

File: src/main/java/com/example/bookstoreapplication/controller/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/categories")
+@Tag(name = "Categories", description = "Category management")
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PostMapping
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapplication/controller/CategoryController.java:32]:
```suggestion
    @PreAuthorize("hasRole('ADMIN')")
```

Comment (d1sam) [reply src/main/java/com/example/bookstoreapplication/controller/CategoryController.java:32]: and in another places

File: src/main/java/com/example/bookstoreapplication/controller/CategoryController.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    @PutMapping("/{id}")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
+    @Operation(summary = "Update category",
+                description = "Updates category by its id(ADMIN only)")
+    public CategoryDto updateCategory(@PathVariable Long id, @RequestBody CategoryRequestDto dto) {
+        return categoryService.update(id, dto);
+    }
+
+    @DeleteMapping("/{id}")
+    @PreAuthorize("hasRole('ROLE_ADMIN')")
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapplication/controller/CategoryController.java:64]:
```suggestion
    @PreAuthorize("hasRole('ROLE_ADMIN')")
    @ResponseStatus(value = HttpStatus.NO_CONTENT)
```

File: src/main/java/com/example/bookstoreapplication/dto/category/CategoryRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Size;
+import lombok.Data;
+
+@Data
+public class CategoryRequestDto {
+    @NotBlank(message = "Category name can not be blank")
+    @Size(min = 5, max = 40, message = "Category name must be from 5 to 40 characters long")
+    private String name;
+    private String description;
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapplication/dto/category/CategoryRequestDto.java:12]: what about max length of description?

File: src/main/java/com/example/bookstoreapplication/mapper/BookMapper.java [hunk 5]
```diff
@@ -13,4 +18,14 @@ public interface BookMapper {
...
     Book toModel(CreateBookRequestDto dto);
+
+    BookDtoWithoutCategoryIds toDtoWithoutCategoryIds(Book book);
+
+    @AfterMapping
+    default void setCategoryIds(@MappingTarget BookDto dto, Book book) {
+        Set<Category> bookCategories = book.getCategories();
+        for (Category category : bookCategories) {
+            dto.getCategoriesIds().add(category.getId());
+        }
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapplication/mapper/BookMapper.java:29]:
```suggestion
          if (book.getCategories() != null) {
      bookDto.setCategoryIds(book.getCategories().stream()
        .map(Category::getId)
        .collect(Collectors.toSet()));
    }
```

Comment (d1sam) [reply src/main/java/com/example/bookstoreapplication/mapper/BookMapper.java:29]: let's use streams where it's possible

File: src/main/java/com/example/bookstoreapplication/model/Book.java [hunk 6]
```diff
@@ -32,4 +36,9 @@ public class Book {
     private String coverImage;
     @Column(nullable = false)
     private boolean isDeleted;
+    @ManyToMany
+    @JoinTable(name = "books_categories",
+            joinColumns = @JoinColumn(name = "book_id"),
+            inverseJoinColumns = @JoinColumn(name = "category_id"))
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapplication/model/Book.java:42]:
```suggestion
            inverseJoinColumns = @JoinColumn(name = "category_id"))
     @ToString.Exclude
  @EqualsAndHashCode.Exclude
```

File: src/main/java/com/example/bookstoreapplication/model/Category.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE categories SET is_deleted = true WHERE id=?")
+@SQLRestriction("is_deleted = false")
+@Table(name = "categories")
+public class Category {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private String name;
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapplication/model/Category.java:24]:
```suggestion
    @Column(nullable = false)
    private String name;
```

File: src/main/java/com/example/bookstoreapplication/model/Category.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    private String description;
+    @Column(nullable = false)
+    private boolean isDeleted;
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapplication/model/Category.java:27]:
```suggestion
    private boolean isDeleted = false;
```

File: src/main/java/com/example/bookstoreapplication/repository/book/BookRepository.java [hunk 9]
```diff
@@ -1,9 +1,10 @@
 package com.example.bookstoreapplication.repository.book;
 
 import com.example.bookstoreapplication.model.Book;
+import java.util.List;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
 
 public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapplication/repository/book/BookRepository.java:8]:
```suggestion
@Repository
public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
```

File: src/main/java/com/example/bookstoreapplication/repository/category/CategoryRepository.java [hunk 10]
```diff
@@ -0,0 +1,8 @@
+package com.example.bookstoreapplication.repository.category;
+
+import com.example.bookstoreapplication.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CategoryRepository extends JpaRepository<Category, Long> {
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapplication/repository/category/CategoryRepository.java:6]:
```suggestion
@Repository
public interface CategoryRepository extends JpaRepository<Category, Long> {
```

File: src/main/resources/db/changelog/changes/09-create-books-categories-table.yaml [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+                    primaryKey: true
+                    foreignKeyName: fr_books_categories_book_id
+                    references: books(id)
+              - column:
+                  name: category_id
+                  type: bigint
+                  constraints:
+                    primaryKey: true
+                    foreignKeyName: fr_books_categories_category_id
+                    references: categories(id)
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/09-create-books-categories-table.yaml:22]:
```suggestion
                    references: categories(id)
                    
```

File: src/main/resources/db/changelog/changes/08-create-categories-table.yaml [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+                    nullable: false
+              - column:
+                  name: description
+                  type: varchar(255)
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Comment (d1sam) [inline src/main/resources/db/changelog/changes/08-create-categories-table.yaml:28]:
```suggestion
                    nullable: false
                    
```

File: src/main/java/com/example/bookstoreapplication/service/category/CategoryServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+@RequiredArgsConstructor
+@Service
+public class CategoryServiceImpl implements CategoryService {
+    private final CategoryRepository repository;
+    private final CategoryMapper mapper;
+
+    @Override
+    public List<CategoryDto> findAll(Pageable pageable) {
+        return repository.findAll().stream().map(mapper::toDto).toList();
```
Comment (d1sam) [inline src/main/java/com/example/bookstoreapplication/service/category/CategoryServiceImpl.java:22]:
```suggestion
        return repository.findAll().stream().map(mapper::toDto).toList();
```
let's map all categories in the `CategoryMapper`

**Review by d1sam — CHANGES_REQUESTED** [review]
Good job, but there are some changes, that should be done according to my comments.

File: src/main/java/com/example/bookstoreapplication/controller/CategoryController.java (hunk 2 above)
Comment (VlasenkoYevhenii) [reply src/main/java/com/example/bookstoreapplication/controller/CategoryController.java:32]: When using the @PreAuthorize annotation with roles, you should match the string that is returned by the getAuthority() method of your Role class.
image

File: src/main/java/com/example/bookstoreapplication/repository/book/BookRepository.java (hunk 9 above)
Comment (VlasenkoYevhenii) [reply src/main/java/com/example/bookstoreapplication/repository/book/BookRepository.java:8]: No need for this annotation as Spring automatically interpreters JpaRepository extending interface as a bean

**Review by andrii-hoienko — APPROVED** [review]
Good job!

**Pull request merged**